extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/tenant/DataSourceTenantConnectionResolver.java

```diff
@@ -25,73 +23,90 @@
-        AgroalDataSource dataSource = tenantDataSource(jpaConfig, tenantId, strategy);
+        AgroalDataSource dataSource = tenantDataSource(dataSourceName, tenantId, multiTenancyStrategy,
+                multiTenancySchemaDataSourceName);
         if (dataSource == null) {
-            throw new IllegalStateException("No instance of datasource found for tenant: " + tenantId);
+            throw new IllegalStateException(
+                    String.format("No instance of datasource found for persistence unit '%1$s' and tenant '%2$s'",
+                            persistenceUnitName, tenantId));
         }
-        if (strategy == MultiTenancyStrategy.SCHEMA) {
-            return new TenantConnectionProvider(tenantId, dataSource);
+        if (multiTenancyStrategy == MultiTenancyStrategy.SCHEMA) {
+            return new SchemaTenantConnectionProvider(tenantId, dataSource);
         }
         return new QuarkusConnectionProvider(dataSource);
     }
 
     /**
      * Create a new data source from the given configuration.
-     * 
+     *
      * @param config Configuration to use.
-     * 
+     *
      * @return New data source instance.
      */
     private static AgroalDataSource createFrom(AgroalDataSourceConfiguration config) {
         try {
             return AgroalDataSource.from(config);
         } catch (SQLException ex) {
-            throw new IllegalStateException("Failed to create a new data source based on the default config", ex);
+            throw new IllegalStateException("Failed to create a new data source based on the existing datasource configuration",
+                    ex);
         }
     }
 
-    /**
-     * Returns either the default data source or the tenant specific one.
-     * 
-     * @param tenantId Tenant identifier. The value is required (non-{@literal null}) in case of
-     *        {@link MultiTenancyStrategy#DATABASE}.
-     * @param strategy Current multitenancy strategy Required value that cannot be {@literal null}.
-     * 
-     * @return Data source.
-     */
-    private static AgroalDataSource tenantDataSource(JPAConfig jpaConfig, String tenantId, MultiTenancyStrategy strategy) {
+    private static AgroalDataSource tenantDataSource(String dataSourceName, String tenantId, MultiTenancyStrategy strategy,
+            String multiTenancySchemaDataSourceName) {
         if (strategy != MultiTenancyStrategy.SCHEMA) {
             return Arc.container().instance(AgroalDataSource.class, new DataSource.DataSourceLiteral(tenantId)).get();
         }
-        String dataSourceName = jpaConfig.getMultiTenancySchemaDataSource();
-        if (dataSourceName == null) {
-            AgroalDataSource dataSource = Arc.container().instance(AgroalDataSource.class).get();
+
+        if (multiTenancySchemaDataSourceName == null) {
+            AgroalDataSource dataSource = getDataSource(dataSourceName);
             return createFrom(dataSource.getConfiguration());
```

Comment:
@michael-schnell @Sanne I'm not sure I understand the need to somehow copy the existing datasource?